### PR TITLE
Add unique index on articles canonical_url

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

Adding the unique index on articles `canonical_url`.

This is in conflict with https://github.com/forem/forem/pull/9788

Thus we need to decide which of these two options:

1. `canonical_url` has to be unique across all articles (drafts and published)
2. `canonical_url` has to be unique only within the scope of published articles (thus this index needs to add `published` in its column list)

I'm not 100% sure what we want here so I though we'd start with this drarft PR

ps. this should be deployed right after https://github.com/forem/forem/pull/10138 - possibly closely after to avoid further duplicates ;)

cc @mstruve @citizen428 due to, respectively, https://github.com/forem/forem/pull/9788#pullrequestreview-472925409 and https://github.com/forem/forem/pull/9788#discussion_r480880463

## Related Tickets & Documents

- https://github.com/forem/forem/pull/8072
- https://github.com/forem/forem/pull/10138

### CODE REVIEW

**Comment by mstruve** [comment]
> 1. canonical_url has to be unique across all articles (drafts and published)

I vote #1 all the way! I think that is the most simple solution and eliminates the problem of someone thinking they have a valid unpublished article and then not being able to publish it.

**Comment by citizen428** [comment]
I think 1. is preferable here, as it's more straight-forward.

File: app/models/article.rb [hunk 1]
```diff
@@ -51,8 +51,8 @@ class Article < ApplicationRecord
 
   validates :body_markdown, length: { minimum: 0, allow_nil: false }, uniqueness: { scope: %i[user_id title] }
   validates :cached_tag_list, length: { maximum: 126 }
-  validates :canonical_url, url: { allow_blank: true, no_local: true,
-                                   schemes: %w[https http] }, uniqueness: { allow_blank: true }
+  validates :canonical_url, uniqueness: { allow_nil: true }
+  validates :canonical_url, url: { allow_blank: true, no_local: true, schemes: %w[https http] }
```
Comment (rhymes) [inline app/models/article.rb:55]: split the validators in 2 lines and replaced `allow_blank` with `allow_nil` to avoid the empty string ;)

**Review by mstruve — APPROVED** [review]

**Comment by rhymes** [comment]
Had to force push to fix the schema conflict

**Review by Zhao-Andy — APPROVED** [review]

**Comment by mstruve** [comment]
Confirmed there are 0 duplicates in the database after the update script

**Pull request merged**